subsys/net/lib/rest_client/src/rest_client.c

```diff
@@ -303,6 +326,11 @@ static int rest_client_do_api_call(struct http_request *http_req,
 		err = -ENOBUFS;
 	} else {
 		err = 0;
+		if (resp_ctx->http_status_code == 0) {
+			LOG_ERR("HTTP null response, i.e., no response received. "
```

Comment:
> it doesn't mean that the server has received all the data

You're technically right -- what it actually means is that the server has received and ACK'ed one or more TCP datagrams, and has decided it doesn't want to hear the rest. This may or may not occur before the entire HTTP request payload is sent to the server. But either way, if you see `http_status_code == 0`, it's because the server sent FIN-ACK with no other response. The only reason I know this is that I'm the author of the PR that added `http_status_code == 0` to Zephyr:

https://github.com/zephyrproject-rtos/zephyr/pull/43453/commits/d31d3bf445c85363db76bf7d8a011cb2a91e9241

If I recall correctly, there isn't any other way for this status code to be generated. The socket has to gracefully close with zero response buffer. Anything else (such as a timeout) will not raise `http_status_code == 0`

Reply:
> In my eyes, `http_status_code == 0` should ideally uniquely represent a FIN-ACK from the TCP endpoint with no response buffer. There shouldn't be any other way for this to happen. If there is, I think we should draft a PR in Zephyr to make this no longer the case.
> 
> I think for most things, like a TCP socket timeout, the HTTP client lib will simply reply with an error.

While socket send errors cause `http_client_req` to fail with error code, any errors during HTTP receive won't get notified. They are just successful responses with the number of bytes sent. HTTP callback won't be called and you will likely just see NULL response. `http_client` library sets errno but in multithread environment, I'm a bit cautious in using it, and rest_client doesn't use that information. We could check errno in REST client and return errors based on that but I'm not sure if this error would be always because of the `http_client_req` call failing.

So, I guess you can get NULL response if socket receive timeout expires, server sends TCP RST, or there is any other socket error as well as this scenario you mentioned about a non-HTTP response.

> If you wind up pursuing this, can you notify me of the PR? I have a vested interest since I recently worked on the http_client regarding `http_status_code == 0`

I'm not planning any Zephyr PR's at the moment as we'll start the holiday season after this week and will be back in August. And fixes through Zephyr would be time-consuming anyway.